[AIRFLOW-4564][AIRFLOW-4523][AIRFLOW-4577] Azure Container Instance  bugfixes and improvements

Trollgeir opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4564
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

- Azure polling errors before container is up fixed
- Azure polling spam in log from TokenRequest fixed
- Only 1 request to Azure per polling iteration (instead of 2)
- Status indication if container fails provisioning
- On-kill procedure (also for execution timeout)
- Environmental variables can be secret/secure (azure)
- Name of container has name-checking at construction time to avoid errors during execution.
- Added GPU support

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- tests/contrib/operators/test_azure_container_instances_operator.py 
  - test_name_checker

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

omusavi commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -104,11 +99,9 @@ class AzureContainerInstancesOperator(BaseOperator):
-
     @apply_defaults
     def __init__(self, ci_conn_id, registry_conn_id, resource_group, name, image, region,
-                 environment_variables=None, volumes=None, memory_in_gb=None, cpu=None,
```
In the original PR for this file, it was requested the None be used as default params:

https://github.com/apache/airflow/pull/4121
https://graysonkoonce.com/always-use-none-for-default-args-in-python/

Trollgeir replied:
Will put Nones back in.

omusavi commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -43,11 +42,6 @@
     ['conn_id', 'account_name', 'share_name', 'mount_path', 'read_only'],
 )
 
-DEFAULT_ENVIRONMENT_VARIABLES = {}  # type: Dict[str, str]
```
Any reason to remove these (including the DEFAULT_MEMORY_IN_GB) instead of inlining as done in the constructor? It reads better with constants, and I believe that was a change made in the original PR

Trollgeir replied:
We could also use `or` at the constructor level.

Should we try to conform with the style of the [kubernetespodoperator](https://github.com/apache/airflow/blob/master/airflow/contrib/operators/kubernetes_pod_operator.py) ?

Trollgeir replied:
Re-implemented.

omusavi commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -208,15 +200,16 @@ def _monitor_logging(self, ci_hook, resource_group, name):
         last_message_logged = None
         last_line_logged = None
-        for _ in range(43200):  # roughly 12 hours
+        while True:
```
instead of while(true), should we instead have a setting for time-out value so DAGs don't just get stuck?

Trollgeir replied:
Will update with a timeout parameter

Trollgeir committed:
[AIRFLOW-4564] ACI bugfixes and improvements

[AIRFLOW-4564] ACI bugfixes and improvements

[AIRFLOW-4564] Updated ACI example

change requests

resolved a merge conflict

[AIRFLOW-4564] ACI hook deprecation updates

small docstring fix

ashb commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -244,7 +321,7 @@ def _monitor_logging(self, ci_hook, resource_group, name):
 
         # no return -> hence still running
-        raise AirflowTaskTimeout("Did not complete on time")
+        raise AirflowTaskTimeout("Did not complete on time (container_timeout = %s)", self.container_timeout)
```
```suggestion
        raise AirflowTaskTimeout("Did not complete in time (container_timeout = %s)", self.container_timeout)
```

Trollgeir replied:
Removed as a part of of the AirflowTaskTimeout procedure.

ashb commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -72,59 +81,96 @@ class AzureContainerInstancesOperator(BaseOperator):
         self.ci_conn_id = ci_conn_id
         self.resource_group = resource_group
-        self.name = name
+        self.name = self._check_name(name)
```
Is it worth checking this again inside execute for the templated case?

Trollgeir replied:
Good idea! Then it can test the template-resolved name (if applicable).

ashb commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -72,59 +81,96 @@ class AzureContainerInstancesOperator(BaseOperator):
+                 command=None,
+                 remove_on_error=True,
+                 fail_if_exists=True,
+                 container_timeout=None,
```
Do we need a new parameter, or could we use the existing execution_timeout from the `BaseOperator`. (Airflow does this:

```python
                        with timeout(int(
                                task_copy.execution_timeout.total_seconds())):
                            result = task_copy.execute(context=context)
                    except AirflowTaskTimeout:
                        task_copy.on_kill()
                        raise
```

So any tidy up can be handed in an on_kill function on the op

Trollgeir replied:
Using your suggestion. Handling the cleanup procedures with the `on_kill()` method.

ashb commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -208,16 +267,30 @@ def _monitor_logging(self, ci_hook, resource_group, name):
         last_line_logged = None
-        for _ in range(43200):  # roughly 12 hours
+
+        for _ in range(self.container_timeout):
```
This is a slightly unusual way of doing a timeout loop. More usual would be something like:

```python
    max_end_time = datetime.now() + self.container_timeount # make it always a timedelta
    while datetime.now() < max_end_time:
        # ...
```

Trollgeir replied:
I initially made it this way to not be dependent on `datetime.now()` beyond init, as there currently is no way of altering the timeout during execution. However this is probably an over-optimization and I don't have any objection to shifting to your suggestion.

Trollgeir replied:
Using AirflowTaskTimeout instead.

ashb commented on airflow/contrib/hooks/azure_container_instance_hook.py:
```diff
@@ -92,43 +92,20 @@ def create_or_update(self, resource_group, name, container_group):
                                                           name,
                                                           container_group)
 
-    def get_state_exitcode_details(self, resource_group, name):
```
Removing this function is a breaking change - whilst the existing operators are _likely_ the only thing callingit would it be possible to add a back-compat shim:

```python
from zope.deprecated import deprecated
 
    @deprecated
    def get_state_exitcode_details(self, resource_group, name):
        c_state = self.get_state(resource_group, name)
        return (c_state.state, c_state.exit_code, c_state.detail_status)
```

(I think. Untested)

Trollgeir replied:
Implemented

ashb commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -154,16 +204,24 @@ def execute(self, context):
 
+            gpu_resource = None
+            if self.gpu:
+                gpu_resource = GpuResource(count=self.gpu[0], sku=self.gpu[1])
```
This is a named tuple, so we can do this right?
```suggestion
                gpu_resource = GpuResource(count=self.gpu.count, sku=self.gpu.sku)
```

Trollgeir replied:
This requires the user to actually import and use the named tuple. Right now the user can choose to use the named tuple or just a tuple.

Should we force the user to import the named tuple? The current examples for the ACI class uses regular tuples.

ashb replied:
Ah, allow a normal tuple for now, but perhaps issue a deprecation warning (in the `__init__` if it is a plain tuple.) and upgrade the object in place?

Trollgeir replied:
If we're going to move away from tuples (implying the user has to import classes), shouldn't we consider using Azure's model object as parameters instead? Right now we're pretty much just forwarding the parameters into the Azure model objects. If in the future Azure adds new functionality into their model object we wouldn't need to upgrade anything on our side.

ashb replied:
Yeah that sounds sensible.

Trollgeir replied:
Using the azure GpuResource model as the parameter type. The operator doesn't currently support gpu, so no need to add a deprecation warning. Those who want to use it should import the model.

ashb commented on airflow/contrib/operators/azure_container_instances_operator.py:
```diff
@@ -154,16 +204,24 @@ def execute(self, context):
+            gpu_resource = None
+            if self.gpu:
+                gpu_resource = GpuResource(count=self.gpu[0], sku=self.gpu[1])
+                self.log.info("GPU count: %.1f, GPU SKU: %s", self.gpu[0], self.gpu[1])
```
```suggestion
                self.log.info("GPU count: %.1f, GPU SKU: %s", self.gpu.count, self.gpu.sku)
```

ashb reviewed (changes requested):
Couple of code changes.

Could you also remove the first commit from this PR -- it's already on master.

ashb commented on airflow/contrib/hooks/azure_container_instance_hook.py:
```diff
@@ -92,43 +92,28 @@ def create_or_update(self, resource_group, name, container_group):
                                                           name,
                                                           container_group)
 
+    from zope.deprecation import deprecation
```
Move to top of file please.

Trollgeir replied:
I noticed it was placed further down here:
https://airflow.apache.org/_modules/airflow/sensors.html

I figured it was placed there so it would be easier to tidy up along with the function in the next update. Should I still move the import?

ashb replied:
Yes please - that is a special case (to not import it unless that setting is enabled), in other places we use it like a normal import, and the Flake8 check will tell us about unused imports once we remove the deprecation.

ashb commented on airflow/contrib/hooks/azure_container_instance_hook.py:
```diff
@@ -92,43 +92,28 @@ def create_or_update(self, resource_group, name, container_group):
-                                                        raw=False)
-        return response.containers[0].instance_view
-
-    def get_messages(self, resource_group, name):
```
Can you add a deprecation shim for this too?

Trollgeir replied:
Will do

ashb reviewed (commented):
Just one last deprecation method to add in please.

Looking good now though
